internal/utils/akka/src/test/java/org/eclipse/ditto/internal/utils/akka/controlflow/TimeMeasuringFlowTest.java

```diff
@@ -74,53 +77,55 @@ public void tearDown() {
     }
 
     @Test
-    @Ignore
-    public void timeMeasuresOnlyTimeOfWrappedFlow() {
+    public void timeMeasuresOnlyTimeOfWrappedFlow() throws Exception {
         final AtomicInteger flowCounter = new AtomicInteger();
-        final Duration sleepDuration = Duration.ofMillis(100);
-        final Flow<String, String, NotUsed> flowThatNeedsSomeTime = Flow.<String, String>fromFunction(x -> {
-            flowCounter.incrementAndGet();
-            TimeUnit.MILLISECONDS.sleep(sleepDuration.toMillis());
-            return x;
-        });
+        final Duration sleepDuration = Duration.ofMillis(1000);
+        final Flow<String, String, NotUsed> flowThatNeedsSomeTime = Flow.<String>create()
+                .delay(sleepDuration, DelayOverflowStrategy.backpressure())
+                .map(x -> {
+                    // Flow.delay is not accurate
+                    flowCounter.incrementAndGet();
+                    return x;
+                });
 
         final PreparedTimer timer = DittoMetrics.timer("test-time-measuring-flow");
         final PreparedTimer timerMock = mock(PreparedTimer.class);
         when(timerMock.start()).thenAnswer(AdditionalAnswers.delegatesTo(timer));
-        final Sink<Duration, TestSubscriber.Probe<Duration>> sink = TestSink.<Duration>probe(system);
-        final Pair<TestSubscriber.Probe<Duration>, Sink<Duration, NotUsed>> sinkPair = sink.preMaterialize(system);
-        final TestSubscriber.Probe<Duration> durationProbe = sinkPair.first();
+
+        final Queue<Duration> durations = new ConcurrentLinkedQueue<>();
         new TestKit(system) {{
+            final int numberOfRepetitions = 10;
             Source.repeat("Test")
-                    .via(TimeMeasuringFlow.measureTimeOf(flowThatNeedsSomeTime, timerMock, sinkPair.second()))
+                    .take(10)
+                    .via(TimeMeasuringFlow.measureTimeOf(flowThatNeedsSomeTime, timerMock, 99, i -> {}, durations::add))
                     .via(flowThatNeedsSomeTime) // This should not influence the time measuring above
                     .to(testSink)
                     .run(system);
 
-            final int numberOfRepetitions = 10;
             final ArrayList<String> expectedResults = new ArrayList<>(numberOfRepetitions);
             for (int i = 0; i < numberOfRepetitions; i++) {
                 expectedResults.add("Test");
             }
-            durationProbe.request(numberOfRepetitions);
             sinkProbe.request(numberOfRepetitions);
             sinkProbe.expectNextN(CollectionConverters.asScala(expectedResults).toSeq());
-            final List<Duration> durations =
-                    CollectionConverters.asJava(durationProbe.expectNextN(numberOfRepetitions));
 
+            if (durations.size() < numberOfRepetitions) {
```

Comment:
Can this case ever happen? Can't we be sure that all expected repetitions are finished, after line 110?

Reply:
Sadly no, because the wiretap has no influence on the base flow. It is possible for the base flow to complete before wiretap handler gets scheduled. A countdown latch is more elegant here. I will amend.